pose_evaluation/metrics/embedding_distance_metric.py

```diff
@@ -1,104 +1,170 @@
+            distance_matrix = self.euclidean_distances(hypotheses, references)
+
+        elif self.kind == "manhattan":
+            distance_matrix = self.manhattan_distances(hypotheses, references)
+
         else:
             raise ValueError(f"Unsupported distance metric: {self.kind}")
 
-        return distance_matrix.cpu()
+        return distance_matrix
+
+    def dot_product(self, hypotheses: list | np.ndarray | Tensor, references: list | np.ndarray | Tensor) -> Tensor:
+        # TODO: test if this gives the same thing as previous matmul implementation, see stack overflow link below:
+        # https://stackoverflow.com/questions/73924697/whats-the-difference-between-torch-mm-torch-matmul-and-torch-mul
+        return st_util.dot_score(hypotheses, references)
+
+    def euclidean_similarities(
+        self, hypotheses: list | np.ndarray | Tensor, references: list | np.ndarray | Tensor
+    ) -> Tensor:
+        """
+        Returns the negative L2 norm/euclidean distances, which is what sentence-transformers uses for similarities.
+        """
+        return st_util.euclidean_sim(hypotheses, references)
+
+    def euclidean_distances(
+        self, hypotheses: list | np.ndarray | Tensor, references: list | np.ndarray | Tensor
+    ) -> Tensor:
+        """
+        Seeing as how sentence-transformers just negates the distances to get "similarities",
+        We can re-negate to get them positive again.
+        """
+        return -self.euclidean_similarities(hypotheses, references)
+
+    def cosine_similarities(
+        self, hypotheses: list | np.ndarray | Tensor, references: list | np.ndarray | Tensor
+    ) -> Tensor:
+        """
+        Calculates cosine similarities, which can be thought of as the angle between two embeddings.
+        The min value is -1 (least similar/pointing directly away), and the max is 1 (exactly the same angle).
+        """
+        return st_util.cos_sim(hypotheses, references)
+
+    def cosine_distances(
+        self, hypotheses: list | np.ndarray | Tensor, references: list | np.ndarray | Tensor
+    ) -> Tensor:
+        """
+        Converts cosine similarities to distances by simply subtracting from 1.
+        Max distance is 2, min distance is 0.
+        """
+        return 1 - self.cosine_similarities(hypotheses, references)
+
+    def manhattan_similarities(
+        self, hypotheses: list | np.ndarray | Tensor, references: list | np.ndarray | Tensor
+    ) -> Tensor:
+        """
+        Get the L1/Manhattan similarities, aka negative distances.
+        """
+        return st_util.manhattan_sim(hypotheses, references)
+
+    def manhattan_distances(
+        self, hypotheses: list | np.ndarray | Tensor, references: list | np.ndarray | Tensor
+    ) -> Tensor:
+        """
+        Sentence transformers defines similarity as negative distances.
+        We can re-negate to recover the distances.
+        """
+        return -self.manhattan_similarities(hypotheses, references)
```

Comment:
I suggest removing all these functions, and instead, calling them directly inside `score_all`. it will reduce the code, and simplify it.

Reply:
Will do!

Reply:
Actually, have a look at the current implementation, which uses a dictionary dispatch style, and let me know what you think